### ISSUE TITLE
Not respecting typeshed METADATA.toml

### ISSUE THREAD

**Issue by ian-r-rose** [microsoft/pyright#1761]
Out-of-date typings for tornado in typeshed have come up a couple of times here (#813, #1662) as well as in `typeshed`. Over in typeshed they were removed and then restored for Python 2 only (as specified in the vendored [`METADATA.toml`](https://github.com/microsoft/pyright/blob/d214e11849853b8f48ac402cd92088e70a74c11c/packages/pyright-internal/typeshed-fallback/stubs/tornado/METADATA.toml#L2-L3)).

As far as I can tell, this metadata isn't checked by pyright, and I'm still seeing these out-of-date stubs getting used. I am not sure where this `METADATA.toml` is specced out (I don't see it in PEP 561), but it seems to be how things are intended to be used in practice.

**To Reproduce**

From a python 3 session:

```python
from tornado.concurrent import Future
Future.__await__  # known missing annotation in old typeshed typings
```

**Expected behavior**

The above should not pull in the old typings as per the `METADATA.toml`

**VS Code extension or command-line**

Command-line, version 1.1.130

**Comment by erictraut** [comment]
The correct fix here is to remove the tornado stubs from pyright and pylance. We don't support Python 2, so there's no reason to include them. Apparently there are another 8 or so third-party stubs that were also marked as "python3 = false" recently. I've removed them as well.

**Comment by ian-r-rose** [comment]
Thanks, looking forward to the next version!

I'm still a little surprised that the module import resolves to the vendored typeshed at all. My reading of the [resolution order docs](https://github.com/microsoft/pyright/blob/master/docs/import-resolution.md#resolution-order) is that the tornado module (which distributes a `py.typed` file) should take precedence over the third-party typeshed. But if I run `pyright --verifytypes tornado` it resolves to the vendored typeshed.

If I manually remove those stubs from the copy of typeshed, it resolves to the expected module with inline types.

**Comment by ian-r-rose** [comment]
To follow up on the above: I think that the behavior of the `ImportResolver` is not quite what is documented, or in PEP 561. Since `tornado` distributes a `py.typed` file, I would expect it to be returned here: https://github.com/microsoft/pyright/blob/5f60965260056e4656ff7eeab0a314318f2b4100/packages/pyright-internal/src/analyzer/importResolver.ts#L1086-L1088 . Instead, control flow continues further in the function, the third-party typeshed is found, and then that is returned. So it correctly finds the import, and then passes by it.

Would it be correct to make the above check something like 
```typescript
 if (thirdPartyImport.isImportFound && (thirdPartyImport.isStubFile || thirdPartyImport.pyTypeInfo)) {
    return thirdPartyImport;
}
```
?

**Comment by erictraut** [comment]
We'd need to discuss this internally before making that change. I agree that the current code doesn't completely conform to the guidance in PEP 561, but there are reasons why we didn't follow PEP 561 exactly. I don't recall whether this is one of those cases or this was just an oversight. It has been a long time since I wrote this code, and it has been modified by others since then.

I think the authors of PEP 561 assumed that "py.typed" libraries were all going to contain complete and high-quality type information, but in practice they often don't.

Are you looking to use a custom (vendored) copy of typeshed? Almost all pyright and pylance users simply use the version of typeshed that we bundle with the tools. We are diligent about keeping this up to date with typeshed. And typeshed is pretty diligent about removing stubs once the library author publishes an updated version that contains inlined types.

**Comment by ian-r-rose** [comment]
Thanks for the context. In any event, removing the python2-only types will fix the issue, but I found the above behavior surprising.

> Are you looking to use a custom (vendored) copy of typeshed? Almost all pyright and pylance users simply use the version of typeshed that we bundle with the tools. We are diligent about keeping this up to date with typeshed. And typeshed is pretty diligent about removing stubs once the library author publishes an updated version that contains inlined types.

I don't really want to use a custom copy of typeshed. In this specific case, the inline types for `tornado` are higher quality than those in typeshed (in fact, I think the authors of tornado were among the driving forces behind PEP 561), and to my mind those should be considered authoritative.

I can say that my experience with this was pretty difficult to debug: 
1. It wasn't obvious that the typeshed annotations were shadowing the "official" `py.typed` ones, and
2. Once I figured out the above, I wasn't able to get pyright to override that without making my own fork of `typeshed`. I don't think I should have to fork `typeshed` to use the officially supported annotations, and it's tough to ask other team members to do that as well.

**Comment by erictraut** [comment]
Those are good points. Thanks for the additional context.

**Comment by erictraut** [comment]
We talked about this briefly, and we think there's a good argument in favor of making a change in the import resolver. We're going to hold off on the change though because it carries with it an element of regression risk, and we are trying to minimize changes over the next few weeks. We'll revisit this again in the not-too-distant future. It would be good to open a new bug report because there are multiple issues covered in this issue.

**Comment by ian-r-rose** [comment]
Okay, thanks for the update, I'll open a more targeted issue

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.131, which I just published. It will also be included in the next release of pylance.

**Comment by ian-r-rose** [comment]
Thanks for the quick fix @erictraut !
